parsers/GT.py

```diff
@@ -1,163 +1,206 @@
-
-DEFAULT_ZONE_KEY = "GT"
+from electricitymap.contrib.lib.models.event_lists import (
+    ProductionBreakdownList,
+    TotalConsumptionList,
+)
+from electricitymap.contrib.lib.models.events import ProductionMix
+from electricitymap.contrib.lib.types import ZoneKey
+from parsers.lib.config import refetch_frequency
+from parsers.lib.exceptions import ParserException
+
+PARSER = "GT.py"
+DEFAULT_ZONE_KEY = ZoneKey("GT")
 TIMEZONE = ZoneInfo("America/Guatemala")
 
-PRODUCTION_THRESHOLD = 10  # MW
-
-DOMAIN = "wl12.amm.org.gt"
-URL = f"https://{DOMAIN}/GraficaPW/graficaCombustible"
+SOURCE = "wl12.amm.org.gt"
+URL = f"https://{SOURCE}/GraficaPW/graficaCombustible"
 
 
 class ApiKind(enum.Enum):
     PRODUCTION = "production"
     CONSUMPTION = "consumption"
 
 
+def _get_api_data(
+    session: Session,
+    target_datetime: datetime,
+    target: ApiKind,
+    num_backlog_days: int,
+) -> dict[datetime, dict]:
+    """Get the one-hourly AMM API data for the desired UTC day + n days of backlog."""
+
+    now_utc = datetime.now(timezone.utc)
+    hour_now_utc = now_utc.replace(minute=0, second=0, microsecond=0)
+
+    target_datetime_utc = target_datetime.astimezone(timezone.utc)
+
+    # The API expects local (TZ) timestamps and can only return one day of data
+    # so might need to do multiple api calls if UTC day straddles multiple local days
+    target_day_utc = datetime.combine(target_datetime_utc, time(), tzinfo=timezone.utc)
+
+    target_day_utc_start = target_day_utc - timedelta(days=num_backlog_days)
+    target_day_utc_end = target_day_utc + timedelta(days=1) - timedelta(microseconds=1)
+
+    local_day_of_target_day_utc_start = datetime.combine(
+        target_day_utc_start.astimezone(TIMEZONE), time(), tzinfo=TIMEZONE
+    )
+    local_day_of_target_day_utc_end = datetime.combine(
+        target_day_utc_end.astimezone(TIMEZONE), time(), tzinfo=TIMEZONE
+    )
+    num_local_days = (
+        local_day_of_target_day_utc_end - local_day_of_target_day_utc_start
+    ).days
+
+    daily_payloads: dict[datetime, list[dict]] = {}
+    for i in range(num_local_days + 1):
+        local_day = local_day_of_target_day_utc_start + timedelta(days=i)
+        response = session.get(URL, params={"dt": local_day.strftime("%d/%m/%Y")})
+        if not response.ok:
+            raise ParserException(
+                PARSER,
+                f"Exception when fetching {target.value} error code: {response.status_code}: {response.text}",
+            )
+
+        # The JSON data returned by the API is a list of objects, each representing one technology type.
+        daily_payloads[local_day] = response.json()
+
+    if not any(daily_payload for daily_payload in daily_payloads.values()):
+        raise ParserException(
+            PARSER,
+            f"Exception when fetching {target.value}: no data available for UTC day {target_day_utc.strftime('%Y-%m-%d %Z')}. "
+            f"Note that historical data is only available for the last 1 year.",
```

Comment:
note: I randomly found this out, but it might be a problem for reproducibility of past results / workflows

Reply:
The most important thing is that the data is available for the last 6 months or so then our automatic refetch pipeline will be able to fill any gaps that might occur if the source is not available during the first parser runs.